flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/BulkSlotProviderImpl.java

```diff
@@ -0,0 +1,271 @@
+		this(slotSelectionStrategy, slotPool, SystemClock.getInstance());
+	}
+
+	@VisibleForTesting
+	BulkSlotProviderImpl(
+			final SlotSelectionStrategy slotSelectionStrategy,
+			final SlotPool slotPool,
+			final Clock clock) {
+
+		this.slotSelectionStrategy = checkNotNull(slotSelectionStrategy);
+		this.slotPool = checkNotNull(slotPool);
+		this.clock = checkNotNull(clock);
+
+		this.slotRequestBulkTracker = new PhysicalSlotRequestBulkTracker(clock);
+
+		this.componentMainThreadExecutor = new ComponentMainThreadExecutor.DummyComponentMainThreadExecutor(
+			"Scheduler is not initialized with proper main thread executor. " +
+				"Call to BulkSlotProvider.start(...) required.");
+	}
+
+	@Override
+	public void start(final ComponentMainThreadExecutor mainThreadExecutor) {
+		this.componentMainThreadExecutor = mainThreadExecutor;
+	}
+
+	@Override
+	public CompletableFuture<Collection<PhysicalSlotRequest.Result>> allocatePhysicalSlots(
+			final Collection<PhysicalSlotRequest> physicalSlotRequests,
+			final Time timeout) {
+
+		componentMainThreadExecutor.assertRunningInMainThread();
+
+		LOG.debug("Received {} slot requests.", physicalSlotRequests.size());
+
+		final PhysicalSlotRequestBulk slotRequestBulk = new PhysicalSlotRequestBulk(physicalSlotRequests);
+
+		final List<CompletableFuture<PhysicalSlotRequest.Result>> resultFutures = new ArrayList<>(physicalSlotRequests.size());
+		for (PhysicalSlotRequest request : physicalSlotRequests) {
+			final CompletableFuture<PhysicalSlotRequest.Result> resultFuture =
+				allocatePhysicalSlot(request, timeout).thenApply(result -> {
+					slotRequestBulk.markRequestFulfilled(
+						result.getSlotRequestId(),
+						result.getPhysicalSlot().getAllocationId());
+
+					return result;
+				});
+			resultFutures.add(resultFuture);
+		}
+
+		slotRequestBulkTracker.track(slotRequestBulk);
+		schedulePendingRequestBulkTimeoutCheck(slotRequestBulk, timeout);
+
+		return FutureUtils.combineAll(resultFutures)
+			.whenComplete((ignore, throwable) -> slotRequestBulkTracker.untrack(slotRequestBulk));
+	}
+
+	private CompletableFuture<PhysicalSlotRequest.Result> allocatePhysicalSlot(
+			final PhysicalSlotRequest physicalSlotRequest,
+			final Time timeout) {
+
+		final SlotRequestId slotRequestId = physicalSlotRequest.getSlotRequestId();
+		final SlotProfile slotProfile = physicalSlotRequest.getSlotProfile();
+		final ResourceProfile resourceProfile = slotProfile.getPhysicalSlotResourceProfile();
+
+		LOG.debug("Received slot request [{}] with resource requirements: {}", slotRequestId, resourceProfile);
+
+		final Optional<PhysicalSlot> availablePhysicalSlot = tryAllocateFromAvailable(slotRequestId, slotProfile);
+
+		final CompletableFuture<PhysicalSlot> slotFuture;
+		if (availablePhysicalSlot.isPresent()) {
+			slotFuture = CompletableFuture.completedFuture(availablePhysicalSlot.get());
+		} else {
+			slotFuture = slotPool.requestNewAllocatedSlotWithoutTimeout(
```

Comment:
maybe no immediate action now, but I find it quite hard to judge about failures in SlotPoolImpl , especially for bulk allocation.
With this new concept of fulfil-ability, maybe we should simplify SlotPoolImpl and let BulkSlotProviderImpl judge about how to interpret external request failures and when to fulfil the requests, it could be a follow-up.
We might want to note that new slot allocation has failed but there are often chances that the pending request can be fulfilled by allocated slots.

Reply:
I'm thinking whether we should treat slot pool more a like a pool, and thus to split concerns of bulk allocation failures and external request failures.
1. bulk allocation fails because it is not fulfillable, we can print debug level logs of desc of slot requests  and state of available/allocated slots in the slot pool.
2. external request failures should to exposed in logs to demonstrate that we fails to extend the pool, which can be the deep reason that why a bulk of slot requests fails. However, external sot request failures does not mean that the internal slot request is definitely to fail.

I'm even thinking to enable auto-retry for failed external sot requests rather than fail the initiating internal slot request to make slot pool more a like a pool. (related to this [comment](https://github.com/apache/flink/pull/12278#discussion_r433607269))

Reply:
Anyhow I totally agreed that we should make it easier for users to debug bulk slot allocation failures. 
Opened a ticket FLINK-18114 to track it.

Reply:
The problem may be in `UnfulfillableSlotRequestException` when there is no point in retrying, see also [comment](https://github.com/apache/flink/pull/12278/commits/7065a71c2911aa3938c827bdb9029dee9268950c#r435362084). It looks this might need more thinking and simplification of design if possible.

Reply:
Agreed. The design should take both "fail-fast on UnfulfillableSlotRequestException" and "not fail streaming slot request if it is fulfillable" into consideration. And related error messages should be enabled for troubleshooting.